File: qnop-ui/src/utils/reviewMailCadence.ts
/*
 * Copyright (c) 2026-present devtank42 GmbH
 *
 * This file is part of qnop (Qualified Notes on Papers).
 *
 * qnop is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * qnop is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with qnop. If not, see <https://www.gnu.org/licenses/>.
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

/**
 * The three cadences for review mail (issue #680), and how to read a stored value.
 *
 * <p>Its own module rather than living beside the profile screen: a file that exports both
 * components and plain functions breaks fast refresh, and this is needed by a test as well.
 */
export const CADENCE_OPTIONS = [
  { value: 'IMMEDIATE', label: 'As it happens' },
  { value: 'DAILY', label: 'Once a day' },
  { value: 'OFF', label: 'Never' },
] as const;

export const CADENCE_HELP: Record<string, string> = {
  IMMEDIATE: 'One email per event.',
  DAILY: 'One summary each morning, in your timezone. Nothing is sent on a quiet day.',
  OFF: 'No review emails. In-app notifications are unaffected.',
};

/**
 * Reads a stored value, tolerating the booleans this setting held before it became three-valued.
 *
 * <p>The mapping matches migration 0035 exactly — `false` was an explicit opt-out and stays `OFF`,
 * `true` becomes `DAILY` — so a browser holding a response from before the migration still shows
 * what the server will actually do.
 */
export function normalizeCadence(stored: string | undefined): string {
  const value = (stored ?? '').trim().toUpperCase();
  if (value === 'TRUE') return 'DAILY';
  if (value === 'FALSE') return 'OFF';
  return CADENCE_OPTIONS.some((option) => option.value === value) ? value : 'DAILY';
}

File: qnop-core/src/main/java/io/qnop/service/notification/ReviewMailCadence.java
/*
 * Copyright (c) 2026-present devtank42 GmbH
 *
 * This file is part of qnop (Qualified Notes on Papers).
 *
 * qnop is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * qnop is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with qnop. If not, see <https://www.gnu.org/licenses/>.
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
package io.qnop.service.notification;

import io.qnop.service.UserSettingKey;
import java.util.Locale;
import java.util.Optional;

/**
 * When a recipient wants review mail (issue #680).
 *
 * <p>Replaces a boolean that offered only "every event" or "nothing". Given that choice on a busy
 * review, the honest answer was nothing — and a notification nobody receives has failed however
 * correct it was.
 *
 * <p>Shared between the mail sink, which sends only for {@link #IMMEDIATE}, and the digest job,
 * which collects for {@link #DAILY}. Both read the same setting, so a recipient cannot end up
 * getting both or neither.
 */
public enum ReviewMailCadence {
  /** One mail per event, as it was before this existed. */
  IMMEDIATE,
  /** One summary per morning, in the recipient's own timezone. */
  DAILY,
  /** No review mail. In-app notifications are unaffected — they are not this setting's business. */
  OFF;

  /**
   * Reads a stored value, tolerating the booleans this setting held before.
   *
   * <p>The legacy mapping matches migration 0035 exactly: {@code false} was an explicit opt-out and
   * stays {@link #OFF}, while {@code true} becomes {@link #DAILY}. Being tolerant here means a
   * deployment whose migration has not run yet still behaves as intended rather than mailing
   * somebody who had said no.
   */
  public static Optional<ReviewMailCadence> parse(String raw) {
    if (raw == null || raw.isBlank()) {
      return Optional.empty();
    }
    String value = raw.trim().toUpperCase(Locale.ROOT);
    return switch (value) {
      case "TRUE" -> Optional.of(DAILY);
      case "FALSE" -> Optional.of(OFF);
      case "IMMEDIATE" -> Optional.of(IMMEDIATE);
      case "DAILY" -> Optional.of(DAILY);
      case "OFF" -> Optional.of(OFF);
      default -> Optional.empty();
    };
  }

  /**
   * What an account that never chose gets — read from the registry rather than repeated here, so
   * the default has exactly one home (ADR-0025).
   */
  public static ReviewMailCadence registryDefault() {
    return parse(UserSettingKey.EMAIL_REVIEW_NOTIFICATIONS.getDefaultValue()).orElse(DAILY);
  }
}
